memote/support/helpers.py

```diff
@@ -777,17 +778,14 @@ def find_bounds(model):
         The metabolic model under investigation.
 
     """
-    bounds = [bound for rxn in model.reactions for bound in rxn.bounds]
     try:
-        lower_bound = np.array(bounds[0::2]).min()
-        upper_bound = np.array(bounds[1::2]).max()
-        lower_bound = -np.amax([np.abs(lower_bound), np.abs(upper_bound)])
-        upper_bound = np.amax([np.abs(lower_bound), np.abs(upper_bound)])
+        lower_bound = np.amin([rxn.lower_bound for rxn in model.reactions])
```

Comment:
Using numpy is often a good idea but in this case you're needlessly creating a list in memory first just to extract the minimum value and then throw away the list. The two versions I previously wrote in comments simply iterate over the bounds. Admittedly the time difference will be miniscule but it's good practice and every little bit adds up.

Reply:
The reason I had to add the list was because if I didn't, I ended up creating generator objects, which do not support `>` or `<` operations, which are used in the functions that call this function.

Reply:
Here let me send you the terminal output of `tox -e py36` after removing the lists to show what I mean.

Reply:
```
(memote-ZVtwQ_6J) bash-3.2$ tox -e py36
py36 develop-inst-nodeps: /Users/sidcha/Documents/projects/memote
py36 installed: alabaster==0.7.10,arrow==0.12.1,asn1crypto==0.24.0,attrs==18.1.0,Babel==2.5.3,binaryornot==0.4.4,cchardet==1.1.3,certifi==2018.4.16,cffi==1.11.5,chardet==3.0.4,click==6.7,click-configfile==0.2.3,click-log==0.2.1,cobra==0.11.3,codecov==2.0.15,configparser==3.5.0,cookiecutter==1.6.0,coverage==4.5.1,cryptography==2.2.2,datapackage==0.8.9,depinfo==1.3.0,docutils==0.14,et-xmlfile==1.0.1,future==0.16.0,gitdb2==2.0.3,GitPython==2.1.9,goodtables==1.0.0,idna==2.6,ijson==2.3,imagesize==1.0.0,importlib-resources==0.5,isodate==0.6.0,jdcal==1.4,Jinja2==2.10,jinja2-ospath==0.2.1,jinja2-time==0.2.0,jsonlines==1.2.0,jsonschema==2.6.0,jsontableschema==0.10.1,linear-tsv==1.1.0,MarkupSafe==1.0,-e git+git@github.com:siddC/memote.git@28e6fa919a7dc1b3ee25659fac1e43141cb50ccb#egg=memote,more-itertools==4.1.0,mpmath==1.0.0,numpy==1.14.3,numpydoc==0.8.0,openpyxl==2.4.11,optlang==1.4.1,packaging==17.1,pandas==0.22.0,pipdeptree==0.12.1,pluggy==0.6.0,poyo==0.4.1,py==1.5.3,pycparser==2.18,PyGithub==1.39,Pygments==2.2.0,PyJWT==1.6.1,pylru==1.1.0,pyparsing==2.2.0,pyperclip==1.6.0,pytest==3.5.1,pytest-cov==2.5.1,pytest-raises==0.7,pytest-rerunfailures==4.0,python-dateutil==2.7.3,python-libsbml==5.16.0,pytz==2018.4,PyYAML==3.12,requests==2.18.4,rfc3986==0.4.1,ruamel.yaml==0.14.12,six==1.11.0,smmap2==2.0.3,snowballstemmer==1.2.1,Sphinx==1.7.4,sphinxcontrib-websupport==1.0.1,SQLAlchemy==1.2.7,swiglpk==1.4.4,sympy==1.1.1,tabulate==0.8.2,tabulator==1.14.0,travis-encrypt==1.1.0,TravisPy==0.3.5,unicodecsv==0.14.1,urllib3==1.22,whichcraft==0.4.1,xlrd==1.1.0,XlsxWriter==1.0.4,xlwt==1.3.0
py36 runtests: PYTHONHASHSEED='2979342377'
py36 runtests: commands[0] | pytest --cov=/Users/sidcha/Documents/projects/memote/memote
========================== test session starts ===========================
platform darwin -- Python 3.6.4, pytest-3.5.1, py-1.5.3, pluggy-0.6.0
rootdir: /Users/sidcha/Documents/projects/memote, inifile: setup.cfg
plugins: rerunfailures-4.0, raises-0.7, cov-2.5.1
collected 374 items                                                      

tests/test_for_utils.py ..........                                 [  2%]
tests/test_for_experimental/test_for_config.py ........            [  4%]
tests/test_for_suite/test_for_api.py ...                           [  5%]
tests/test_for_support/test_for_annotation.py .................... [ 10%]
........................................................           [ 25%]
tests/test_for_support/test_for_basic.py ..............FFF.......F [ 32%]
F...........F.FF.........                                          [ 39%]
tests/test_for_support/test_for_biomass.py ....................... [ 45%]
........                                                           [ 47%]
tests/test_for_support/test_for_consistency.py ................... [ 52%]
.........................................                          [ 63%]
tests/test_for_support/test_for_helpers.py ....................... [ 69%]
................                                                   [ 74%]
tests/test_for_support/test_for_matrix.py ....................     [ 79%]
tests/test_for_support/test_for_sbo.py ......                      [ 81%]
tests/test_for_support/test_for_syntax.py ........................ [ 87%]
..                                                                 [ 87%]
tests/test_for_experimental/test_for_experimental_base.py ........ [ 90%]
.....                                                              [ 91%]
tests/test_for_experimental/test_for_medium.py .....               [ 92%]
tests/test_for_experimental/test_for_tabular.py ........           [ 94%]
tests/test_for_suite/test_for_api.py xx                            [ 95%]
tests/test_for_suite/test_for_collect.py .                         [ 95%]
tests/test_for_suite/test_for_cli/test_for_callbacks.py .          [ 95%]
tests/test_for_suite/test_for_cli/test_for_reports.py .            [ 96%]
tests/test_for_suite/test_for_cli/test_for_runner.py .             [ 96%]
tests/test_for_suite/test_for_cli/test_for_reports.py .            [ 96%]
tests/test_for_suite/test_for_cli/test_for_runner.py .s            [ 97%]
tests/test_for_suite/test_for_results/test_for_models.py .......   [ 99%]
tests/test_for_support/test_for_helpers.py ...                     [100%]

================================ FAILURES ================================
___ test_find_nonzero_constrained_reactions[glpk-unconstrained_rxn-0] ____

model = <Model unconstrained_rxn at 0x110e7ec88>, num = 0

    @pytest.mark.parametrize("model, num", [
        ("empty", 0),
        ("unconstrained_rxn", 0),
        ("nonzero_constrained_rxn", 1),
        ("no_nonzero_constrained_rxn", 0),
    ], indirect=["model"])
    def test_find_nonzero_constrained_reactions(model, num):
        """Expect amount of non-zero rxns to be identified correctly."""
>       nonzero_constrained_rxns = basic.find_nonzero_constrained_reactions(model)

tests/test_for_support/test_for_basic.py:429: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
memote/support/basic.py:52: in find_nonzero_constrained_reactions
    return [rxn for rxn in model.reactions if
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <list_iterator object at 0x1101e6400>

    return [rxn for rxn in model.reactions if
>           0 > rxn.lower_bound > lower_bound or
            0 < rxn.upper_bound < upper_bound]
E   TypeError: '>' not supported between instances of 'int' and 'generator'

memote/support/basic.py:53: TypeError
 test_find_nonzero_constrained_reactions[glpk-nonzero_constrained_rxn-1] _

model = <Model nonzero_constrained_rxn at 0x110efe0b8>, num = 1

    @pytest.mark.parametrize("model, num", [
        ("empty", 0),
        ("unconstrained_rxn", 0),
        ("nonzero_constrained_rxn", 1),
        ("no_nonzero_constrained_rxn", 0),
    ], indirect=["model"])
    def test_find_nonzero_constrained_reactions(model, num):
        """Expect amount of non-zero rxns to be identified correctly."""
>       nonzero_constrained_rxns = basic.find_nonzero_constrained_reactions(model)

tests/test_for_support/test_for_basic.py:429: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
memote/support/basic.py:52: in find_nonzero_constrained_reactions
    return [rxn for rxn in model.reactions if
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <list_iterator object at 0x110369a58>

    return [rxn for rxn in model.reactions if
>           0 > rxn.lower_bound > lower_bound or
            0 < rxn.upper_bound < upper_bound]
E   TypeError: '>' not supported between instances of 'int' and 'generator'

memote/support/basic.py:53: TypeError
 test_find_nonzero_constrained_reactions[glpk-no_nonzero_constrained_rxn-0] 

model = <Model no_nonzero_constrained_rxn at 0x110ee7eb8>, num = 0

    @pytest.mark.parametrize("model, num", [
        ("empty", 0),
        ("unconstrained_rxn", 0),
        ("nonzero_constrained_rxn", 1),
        ("no_nonzero_constrained_rxn", 0),
    ], indirect=["model"])
    def test_find_nonzero_constrained_reactions(model, num):
        """Expect amount of non-zero rxns to be identified correctly."""
>       nonzero_constrained_rxns = basic.find_nonzero_constrained_reactions(model)

tests/test_for_support/test_for_basic.py:429: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
memote/support/basic.py:52: in find_nonzero_constrained_reactions
    return [rxn for rxn in model.reactions if
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <list_iterator object at 0x110e8d390>

    return [rxn for rxn in model.reactions if
>           0 > rxn.lower_bound > lower_bound or
            0 < rxn.upper_bound < upper_bound]
E   TypeError: '>' not supported between instances of 'int' and 'generator'

memote/support/basic.py:53: TypeError
______ test_find_unconstrained_reactions[glpk-unconstrained_rxn-1] _______

model = <Model unconstrained_rxn at 0x110ed11d0>, num = 1

    @pytest.mark.parametrize("model, num", [
        ("empty", 0),
        ("unconstrained_rxn", 1),
        ("zero_constrained_rxn", 0),
    ], indirect=["model"])
    def test_find_unconstrained_reactions(model, num):
        """Expect amount of unconstrained rxns to be identified correctly."""
>       unconstrained_rxns = basic.find_unconstrained_reactions(model)

tests/test_for_support/test_for_basic.py:462: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
memote/support/basic.py:72: in find_unconstrained_reactions
    return [rxn for rxn in model.reactions if
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <list_iterator object at 0x110efef60>

    return [rxn for rxn in model.reactions if
>           rxn.lower_bound <= lower_bound and
            rxn.upper_bound >= upper_bound]
E   TypeError: '<=' not supported between instances of 'int' and 'generator'

memote/support/basic.py:73: TypeError
_____ test_find_unconstrained_reactions[glpk-zero_constrained_rxn-0] _____

model = <Model zero_constrained_rxn at 0x110026d68>, num = 0

    @pytest.mark.parametrize("model, num", [
        ("empty", 0),
        ("unconstrained_rxn", 1),
        ("zero_constrained_rxn", 0),
    ], indirect=["model"])
    def test_find_unconstrained_reactions(model, num):
        """Expect amount of unconstrained rxns to be identified correctly."""
>       unconstrained_rxns = basic.find_unconstrained_reactions(model)

tests/test_for_support/test_for_basic.py:462: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
memote/support/basic.py:72: in find_unconstrained_reactions
    return [rxn for rxn in model.reactions if
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <list_iterator object at 0x10ff39940>

    return [rxn for rxn in model.reactions if
>           rxn.lower_bound <= lower_bound and
            rxn.upper_bound >= upper_bound]
E   TypeError: '<=' not supported between instances of 'int' and 'generator'

memote/support/basic.py:73: TypeError
 test_find_constrained_pure_metabolic_reactions[glpk-ngam_and_atpsynthase-0] 

model = <Model ngam_and_atpsynthase at 0x110ed2828>, num = 0

    @pytest.mark.parametrize("model, num", [
        ("ngam_present", 1),
        ("ngam_and_atpsynthase", 0),
        ("non_metabolic_reactions", 0)
    ], indirect=["model"])
    def test_find_constrained_pure_metabolic_reactions(model, num):
        """Expect num of contrained metabolic rxns to be identified correctly."""
        pmr = basic.find_pure_metabolic_reactions(model)
        contrained_pmr = set(
>           [rxn for rxn in pmr if basic.is_constrained_reaction(model, rxn)])

tests/test_for_support/test_for_basic.py:516: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_for_support/test_for_basic.py:516: in <listcomp>
    [rxn for rxn in pmr if basic.is_constrained_reaction(model, rxn)])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = <Model ngam_and_atpsynthase at 0x110ed2828>
rxn = <Reaction NGAM at 0x110ec1b00>

    def is_constrained_reaction(model, rxn):
        """Return whether a reaction has fixed constraints."""
        lower_bound, upper_bound = helpers.find_bounds(model)
        if rxn.reversibility:
            return rxn.lower_bound > lower_bound or rxn.upper_bound < upper_bound
        else:
>           return rxn.lower_bound > 0 or rxn.upper_bound < upper_bound
E           TypeError: '<' not supported between instances of 'int' and 'generator'

memote/support/basic.py:232: TypeError
 test_find_constrained_transport_reactions[glpk-transport_gpr_constrained-1] 

model = <Model transport_gpr_constrained at 0x110edb5c0>, num = 1

    @pytest.mark.parametrize("model, num", [
        ("transport_gpr_constrained", 1),
        ("transport_gpr", 0),
        ("ngam_and_atpsynthase", 0)
    ], indirect=["model"])
    def test_find_constrained_transport_reactions(model, num):
        """Expect num of contrained transport rxns to be identified correctly."""
        transporters = helpers.find_transport_reactions(model)
        constrained_transporters = set(
>           [rxn for rxn in transporters if basic.is_constrained_reaction(
                model, rxn)])

tests/test_for_support/test_for_basic.py:529: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_for_support/test_for_basic.py:530: in <listcomp>
    model, rxn)])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = <Model transport_gpr_constrained at 0x110edb5c0>
rxn = <Reaction ANTI at 0x110edba20>

    def is_constrained_reaction(model, rxn):
        """Return whether a reaction has fixed constraints."""
        lower_bound, upper_bound = helpers.find_bounds(model)
        if rxn.reversibility:
            return rxn.lower_bound > lower_bound or rxn.upper_bound < upper_bound
        else:
>           return rxn.lower_bound > 0 or rxn.upper_bound < upper_bound
E           TypeError: '<' not supported between instances of 'float' and 'generator'

memote/support/basic.py:232: TypeError
____ test_find_constrained_transport_reactions[glpk-transport_gpr-0] _____

model = <Model transport_gpr at 0x110ed2d68>, num = 0

    @pytest.mark.parametrize("model, num", [
        ("transport_gpr_constrained", 1),
        ("transport_gpr", 0),
        ("ngam_and_atpsynthase", 0)
    ], indirect=["model"])
    def test_find_constrained_transport_reactions(model, num):
        """Expect num of contrained transport rxns to be identified correctly."""
        transporters = helpers.find_transport_reactions(model)
        constrained_transporters = set(
>           [rxn for rxn in transporters if basic.is_constrained_reaction(
                model, rxn)])

tests/test_for_support/test_for_basic.py:529: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_for_support/test_for_basic.py:530: in <listcomp>
    model, rxn)])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = <Model transport_gpr at 0x110ed2d68>
rxn = <Reaction UNI at 0x111075208>

    def is_constrained_reaction(model, rxn):
        """Return whether a reaction has fixed constraints."""
        lower_bound, upper_bound = helpers.find_bounds(model)
        if rxn.reversibility:
            return rxn.lower_bound > lower_bound or rxn.upper_bound < upper_bound
        else:
>           return rxn.lower_bound > 0 or rxn.upper_bound < upper_bound
E           TypeError: '<' not supported between instances of 'float' and 'generator'

memote/support/basic.py:232: TypeError

---------- coverage: platform darwin, python 3.6.4-final-0 -----------
Name                                       Stmts   Miss Branch BrPart  Cover
----------------------------------------------------------------------------
memote/__init__.py                             8      0      0      0   100%
memote/experimental/__init__.py                3      0      0      0   100%
memote/experimental/checks.py                 26     15     12      0    29%
memote/experimental/config.py                107     18     46     15    77%
memote/experimental/essentiality.py           28      2      8      2    89%
memote/experimental/experiment.py             13      0      0      0   100%
memote/experimental/experimental_base.py      36      5      8      1    77%
memote/experimental/growth.py                 12      1      0      0    92%
memote/experimental/medium.py                 24      1     10      2    91%
memote/experimental/schemata/__init__.py       2      0      0      0   100%
memote/experimental/tabular.py                14      1      6      1    90%
memote/support/__init__.py                     2      0      0      0   100%
memote/support/annotation.py                  40      0     18      1    98%
memote/support/basic.py                       69      1     47      2    97%
memote/support/biomass.py                    124      2     56      3    97%
memote/support/consistency.py                198     19     88      4    89%
memote/support/consistency_helpers.py         91      3     32      0    98%
memote/support/data/__init__.py                2      0      0      0   100%
memote/support/essentiality.py                36     30      0      0    17%
memote/support/helpers.py                    195     13    133      5    94%
memote/support/matrix.py                      26      0      0      0   100%
memote/support/sbo.py                          8      0      4      0   100%
memote/support/syntax.py                      74      0     34      2    98%
memote/utils.py                               51     11     16      0    72%
----------------------------------------------------------------------------
TOTAL                                       1189    122    518     38    89%

============================ warnings summary ============================
tests/test_for_suite/test_for_api.py::test_snapshot_report_file[glpk-complete_failure]
  /Users/sidcha/Documents/projects/memote/memote/suite/reporting/config.py:57 UnsafeLoaderWarning: 
  The default 'Loader' for 'load(stream)' without further arguments can be unsafe.
  Use 'load(stream, Loader=ruamel.yaml.Loader)' explicitly if that is OK.
  Alternatively include the following in your code:
  
    import warnings
    warnings.simplefilter('ignore', ruamel.yaml.error.UnsafeLoaderWarning)
  
  In most other cases you should consider using 'safe_load(stream)'

tests/test_for_suite/test_for_cli/test_for_reports.py::test_snapshot[ecoli-core]
  /Users/sidcha/Documents/projects/memote/memote/suite/reporting/config.py:57 UnsafeLoaderWarning: 
  The default 'Loader' for 'load(stream)' without further arguments can be unsafe.
  Use 'load(stream, Loader=ruamel.yaml.Loader)' explicitly if that is OK.
  Alternatively include the following in your code:
  
    import warnings
    warnings.simplefilter('ignore', ruamel.yaml.error.UnsafeLoaderWarning)
  
  In most other cases you should consider using 'safe_load(stream)'

-- Docs: http://doc.pytest.org/en/latest/warnings.html
 8 failed, 363 passed, 1 skipped, 2 xfailed, 2 warnings in 30.55 seconds =
ERROR: InvocationError: '/Users/sidcha/Documents/projects/memote/.tox/py36/bin/pytest --cov=/Users/sidcha/Documents/projects/memote/memote'
________________________________ summary _________________________________
ERROR:   py36: commands failed
(memote-ZVtwQ_6J) bash-3.2$ 

```

Reply:
^ removing only `[` and `]` gives the above `TypeError` exceptions warning about comparison between `int` and `generator` objects.

Reply:
Did you switch to the native Python `min` and `max` rather than the numpy versions, too?

Reply:
No, it’s still kept as `np.amin` and `np.amax`

Reply:
Should I change it to `min` and `max`?

Reply:
Yes, please give it a try. They should return a single number. The numpy functions cannot deal with generator statements so they simply pass it on.

Reply:
Okay yeah, that passes `tox -e py36` with no errors! I'm committing the change now! However, when I ran it the test the first time I got a weird `InvoctationError`. I tried it once more time and it ran without a problem, though. I'll send you the the terminal output on Slack.